app/controllers/users.js

```diff
@@ -168,27 +167,35 @@ exports.user = function (req, res, next, id) {
     });
 };
 
-exports.signupWithEmail = function (req, res) {
-  // get the user credentials from form  req.body.password
-  // req.body.email
-  User
-    .findOne({ email: req.body.email })
-    .then((existingUser, err) => {
-      if (err) throw err;
-      if (existingUser) {
-        return res.json({ message: 'A user with this email address already exists' });
+exports.signUp = (req, res) => {
+  if (req.body.name && req.body.password && req.body.email) {
+    User.findOne({
+      email: req.body.email
+    }).exec((err, existingUser) => {
+      if (!existingUser) {
+        const user = new User(req.body);
+        // Switch the user's avatar index to an actual avatar url
+        user.avatar = avatars[user.avatar];
+        user.provider = 'jwt';
+        user.save((err) => {
+          if (err) {
+            return res.status(400).json({ message: 'Could not create user' });
```

Comment:
The status code here should be 500 not 400, and you might need to return the actual error rather than your custom one.

```js
res.status(500).json(err)
```